test/stream_decoder_test.cc

```diff
@@ -167,13 +170,61 @@ TEST_F(StreamDecoderTest, LatencyIsMeasured) {
   EXPECT_EQ(1, stream_decoder_export_latency_callbacks_);
 }
 
+TEST_F(StreamDecoderTest, EmptyRequestBodyWithNonZeroRequestBodySize) {
+  std::string expected_body = "aaaa";
+  Envoy::Buffer::OwnedImpl buf(expected_body);
+  auto decoder = new StreamDecoder(
+      *dispatcher_, time_system_, *this, [](bool, bool) {}, connect_statistic_, latency_statistic_,
+      response_header_size_statistic_, response_body_size_statistic_, origin_latency_statistic_,
+      request_headers_, "", false, 4, random_generator_, tracer_, "");
+  Envoy::Http::MockRequestEncoder stream_encoder;
+  EXPECT_CALL(stream_encoder, getStream());
+  Envoy::Upstream::HostDescriptionConstSharedPtr ptr;
+  NiceMock<Envoy::StreamInfo::MockStreamInfo> stream_info;
+  Envoy::Buffer::OwnedImpl captured_encoder_body_input_buf;
+  EXPECT_CALL(stream_encoder,
+              encodeHeaders(Envoy::HeaderMapEqualRef(request_headers_.get()), false));
+  EXPECT_CALL(stream_encoder, encodeData(_, true))
+      .Times(1)
+      .WillOnce(testing::DoAll(SaveArg<0>(&captured_encoder_body_input_buf)));
+  decoder->onPoolReady(stream_encoder, ptr, stream_info,
+                       {} /*absl::optional<Envoy::Http::Protocol> protocol*/);
+  decoder->decodeHeaders(std::move(test_header_), false);
+  EXPECT_EQ(captured_encoder_body_input_buf.toString(), expected_body);
+  delete decoder;
+}
+
+TEST_F(StreamDecoderTest, NonEmptyRequestBodyWithNonZeroRequestBodySize) {
+  std::string json_body = R"({"Message": "Hello"})";
+  Envoy::Buffer::OwnedImpl json_buf(json_body);
+  auto decoder = new StreamDecoder(
+      *dispatcher_, time_system_, *this, [](bool, bool) {}, connect_statistic_, latency_statistic_,
+      response_header_size_statistic_, response_body_size_statistic_, origin_latency_statistic_,
+      request_headers_, json_body, false, 20, random_generator_, tracer_, "");
```

Comment:
IIUC the 20 can now become 0 and it should still work, because the behavior is determined by json_body being nonempty.

Reply:
+1 we better set it to 0.